### DISCUSSION TITLE
Consistent background behavior

### DISCUSSION THREAD

**Discussion by strider72** [chillerlan/php-qrcode#199] [Ideas]
Backgrounds behave differently with SVGs compared to other common image formats. Even if you ask for solid background, its not there with SVG. Dev has stated this is because you can style it with CSS anyway, but this doesn't apply if, for example, you're generating an image for download. Even with CSS, there is no obvious element to style.  I'm working with circular modules, and even with "visible" background the background itself is also dots with transparent gaps.

## Feature request

With SVG format, please:

* add a containing square that is behind all other elements, sized to the outer edge of the Quiet Zone, and
* if image is not transparent, set the box to the background color. If image is transparent, box is either transparent or just `display:none`.

A simple box to style would simplify things greatly. The box following the requested background formatting/color is even better.

**Comment by codemasher** [comment]
The thing about SVG is that unlike solid colors in raster graphics it has not only opaque fills via the `fill` attribute but it can also support gradients via CSS or inline `style` attribute or `defs` and opacity via the `opacity` attribute or CSS or... I could add a "simple" solution that may support one specific case, but any other case would still require to extend the SVG output class - which you can do also, specific to your use case - which is why i made the library extensible in first place. 
So what you can do for your specific case is to extend the class and add a `fill` or `style` attribute to the header over here:
https://github.com/chillerlan/php-qrcode/blob/c8c37a4e36300f1fc4a7fb44ee22186eac1e60a4/src/Output/QRMarkupSVG.php#L75-L90

**Comment by strider72** [comment]
SVG has a lot of flexibility, but can you agree that people might expect the image to be the same, given the same parameters?  I mean, a solid color (probably white) background for a QR Code is an **incredibly common** use case.  It makes sense that if somebody sets background color (and transparency false) that it would have a background color.  Indeed it works exactly like that with other formats.

If somebody *wants* a gradient or something unusual like that, they can still extend the class. That's a bonus of using the flexible SVG format.  But for the very common case that many users might want, it makes sense to make the already-existing one-liner option *work*.

~~I could extend the header to add a `rect` element, but to do so I have to duplicate the code of the entire `header` function.  As structured, I can't just run `PARENT::header()`, because the insertion happens in the middle, I have to manually reproduce the whole function. Now, any time the library is updated I need to check that you didn't change something else. Something as minor as a white background has now become an ongoing maintenance task.~~

~~(Yeah, I could use PHP to parse the header XML structure, but... no. Please God, no.)~~

[EDIT: I was wrong on that point. Just use parent and add to the end.  Please see [this comment](https://github.com/chillerlan/php-qrcode/discussions/199#discussioncomment-5736234).  I think the rest regarding basic vs advanced usage still stands.]

Heck, just adding something like the following line before the first path would be helpful:

```
<rect width="%2$s" height="%2$s" class="qr_BG" />
```

(It wouldn't help my "make a file for download" situation, but it would be helpful for easy CSS styling.)

Best would be that *if* transparency is false, and *if* a background color is set, add

```
<rect width="%2$s" height="%2$s" class="qr_BG" style="fill:rgb( BACKGROUND COLOR );" />
```

If somebody is doing something more complex such as a gradient, they can extend the class as needed and ignore the "basic" background options.  So doing this doesn't *hurt* the people doing advanced stuff, but it greatly simplifies things for people doing basic stuff.

**Reply by codemasher** [reply]
> SVG has a lot of flexibility, but can you agree that people might expect the image to be the same, given the same parameters?

Unlike the raster image formats (or PDF and EPS for that matter), SVG is used almost exclusively in web/html context where it can be easily manipulated via CSS and be wrapped in containers. While consistent behaviour might be an argument, I don't think there is *the one* good solution to implement it, nor enough reason to do so since there are countless other ways to solve it externally. So I'll leave it to the implementer.

**Reply by strider72** [reply]
> > SVG has a lot of flexibility, but can you agree that people might expect the image to be the same, given the same parameters?
> 
> Unlike the raster image formats (or PDF and EPS for that matter), SVG is used almost exclusively in web/html context where it can be easily manipulated via CSS and be wrapped in containers. While consistent behaviour might be an argument, I don't think there is _the one_ good solution to implement it, nor enough reason to do so since there are countless other ways to solve it externally. So I'll leave it to the implementer.

I use it wherever possible because I don't have to worry about resolution.  I (for example) see our corporate logo on documents all the time where the user had some cruddy low resolution copy of the logo and used that. SVG solves the issue, and I've been working to get SVG logo in people's hands for the sake of brand consistency.  Same with QR codes, especially with round modules -- no chunky circles.  (Admittedly not a big deal if everything is a square.)

As a coder, whenever I'm deciding whether to add a feature, I have a couple considerations.

1. **Effort vs. return**. How hard is it to implement, balanced against the benefit. (Benefit includes how common the use case is -- how many users are affected? Also, how much work am I preventing.)  *In this case, implementation is one or two lines of code. Very easy to add to core code. Conversely, for the end user to add a simple white background requires multiple custom classes, plus time going through documentation figuring out exactly what they need to do.*

2. **Consistency**. Am I making it work in an expected or random (weird) way. *If the system allows users to set a background, they will expect a background.  If the background setting does nothing, they will waste time puzzling over why "the code doesn't work" -- ***it looks like a bug***.*

3. **Does it create limitations?**  Do I end up *requiring* things to be done in a certain way that limits other options?  *In this case, no. It adds expected "basic" behavior, but users with advanced needs can still do it the more complicated way.*

I'll finish with this: You said it can be done with simple CSS styling. Unless I'm missing something, I don't believe you're correct on that. (I'm very good with CSS, but not super experienced with SVG.) My first request was to simply add the `<rect>` to the SVG header so that I *could* style it.  Without it, there's nothing to pin the style on.  Just the `<rect>` wouldn't solve my issue of stand alone-SVG QR codes with background, but would, as you say, make it easily styleable when using on a web page.

Hey... this is Github after all.  If I submitted a good patch that doesn't break anything, would you consider adding it?

**Reply by codemasher** [reply]
> You said it can be done with simple CSS styling. Unless I'm missing something, I don't believe you're correct on that. 

This works perfectly fine (either in the svg defs or via external css):
```css
svg{
    background: green;
}
```
as well as this would
```html
<img src="[base64 qrcode]" style="background: green;" />
```
Do you want me to post more examples?
Also i don't understand what's so hard about extending the `QRMarkupSVG` class as it is shown in the examples and add the `rect` you want? Just... do it?

```php
class MySVG extends QRMarkupSVG{

	/**
	 * @inheritDoc
	 */
	protected function header():string{
		$width  = ($this->options->svgWidth !== null) ? sprintf(' width="%s"', $this->options->svgWidth) : '';
		$height = ($this->options->svgHeight !== null) ? sprintf(' height="%s"', $this->options->svgHeight) : '';

		/** @noinspection HtmlUnknownAttribute */
		return sprintf(
			'<?xml version="1.0" encoding="UTF-8"?>%6$s'.
			'<svg xmlns="http://www.w3.org/2000/svg" class="qr-svg %1$s" viewBox="0 0 %2$s %2$s" preserveAspectRatio="%3$s"%4$s%5$s>%6$s'.
			'<rect width="%2$s" height="%2$s" class="qr_BG" style="fill:rgb( BACKGROUND COLOR );" />%6$s',
			$this->options->cssClass,
			($this->options->svgViewBoxSize ?? $this->moduleCount),
			$this->options->svgPreserveAspectRatio,
			$width,
			$height,
			$this->options->eol
		);
	}

}
```

**Reply by strider72** [reply]
> ```css
> svg{
>     background: green;
> }
> ```

Weird. This did not work for me. First thing I tried.  Something unrelated must be messing it up.

(As I said, not super experienced with SVG styling. I've styled paths and `<g>`s, but generally not the svg element directly.)

**Reply by strider72** [reply]
> Also i don't understand what's so hard about extending the `QRMarkupSVG` class as it is shown in the examples and add the `rect` you want? Just... do it?

A reasonable question.  The answer is that I puzzled over this problem for a long time. Hours. It was *maddening* trying to figure out why the heck this basic functionality *didn't just work*.  I want to spare every other user you have this frustration.  Set background, background works, done.

You're deliberately designing in a weird exception to the rule.  "Here's how to set a background color. Oh, except if you're using this file format -- then you have to program two custom classes...."  If this were some rare outlying use case, I could see it, but a plain background for a QR code seems like the most common use case I can imagine -- far more common than gradients, or whatever else funky styling somebody might want; and yes, even more common than transparent because you want the thing to be scannable.

You're so close to the code that it all seems super obvious to you. It's not obvious at all to some random person just looking to generate some QRs.

And again, adding it doesn't prevent the advanced usage. Doesn't make it the least bit harder. *There's no downside.*  And SVG isn't just for advanced users -- it's also the clear choice for people who simply want the image to be resolution-agnostic.

Honestly at this point I've already coded the custom classes.  But overall this design choice cost me hours of wasted time. Potentially multiply that by *every user you have* -- or even just a good number of them.  That's why I'm fighting for this change.  It's one person adding a line or two of code, vs. many people struggling to figure out basic functionality. It seems obvious to me.

[EDIT: At any rate... I won't keep hounding you about it.  I've argued the case as best I can.  It's a good package you've made.  Incidentally, I also made a class for using your QRs directly within the RobThree 2FA/TOTP library, which I've posted over at that project.]

**Reply by codemasher** [reply]
> You're deliberately designing in a weird exception to the rule.

There is no rule, only exceptions. Every output class works differently because of either limitations or abundant ways to solve things. And SVG just has too many ways of solving the thing you want and your proposed "solution" is rather limited to one specific use case, which is exactly why i made this library extensible - i even went so far and extracted the methods for every single part of the SVG to make changes as simple as possible. If that isn't enough for you, i can't help you.

**Reply by strider72** [reply]
> There is no rule

Sure there is. If you want to set a background color, set `bgColor`, and make sure `imageTransparent` is false.  It's readily understandable and quite clear.

It's not the difficulty of typing a certain number of characters. It's the difficulty of figuring out why setting a background color *doesn't work as expected*.  Users won't see it as a deliberate design decision; they'll see it as a *bug*.

There are a lot of classes in this package. Setting a custom `outputType` and `outputInterface` is not obvious to somebody new to the package -- especially when there exists an obvious `bgColor` setting that doesn't work.   So again, it's the issue of sparing basic users having to figure out the complexities; not the physical difficulty of implementing it after you've figured it all out.

**Reply by strider72** [reply]
Let me respond another way:  You asked me before why I don't just do it, and I believe I identified good reasons to add it to "core".  I'm now asking you the same thing. What are the reasons to *not* have `bgcolor` work for SVG?

It doesn't stop advanced users from extending the classes as they can now; but it greatly simplifies things for all the people who *aren't* doing that.

**Reply by codemasher** [reply]
> Setting a custom outputType and outputInterface is not obvious to somebody new to the package

There is literally [an example](https://github.com/chillerlan/php-qrcode/blob/main/examples/custom_output.php) for it? That sounds a lot like a you problem.
Not to mention i added the bgcolor option the last time when you were here for a similar "issue". Back then it was an option exclusive to the GD output class but i realized that Imagick handles it similar and it could be used there as well, which is also stated in the description.

Just to make it clear: i will not "solve" a non-issue just for the sake of consistency (or what you perceive as that). The average user will put the SVG in either a `<div>` or `<img>` tag and add a background property to these containers or they will apply a css rule directly as i've shown above. If you want an extra for your specific case, you will need to walk the extra mile - it's that simple.

**Reply by strider72** [reply]
>>is it possible to generate a QR-Code with a white background?

>Hey! Sure there is an option for it - just set QROptions::$imageTransparent to false et voilà!

https://github.com/chillerlan/php-qrcode/issues/55

**Reply by codemasher** [reply]
What's the gotcha here? GD output was the default until v4.x and the `image*` prefixed options used to be exclusive to that.

**Reply by strider72** [reply]
No gotcha. Just... your answer there was great because *the solution is so simple and it works*. That's really good design.

And yes, it's better than if your answer had been "Sure... just look through the documentation and find the custom output examples; then write some custom classes, and....". If they guy has asked how to make gradient backgrounds, then "here's an example" is perfect.  But an obvious, simple use case had a clear concise answer here.

There was a reason to add the bgcolor option -- (or more specifically, the `imageTransparent = false` option?). It's straightforward and useful.

> i added the bgcolor option the last time when you were here for a similar "issue"

You did??  I'm not sure if that was me.  I've asked a couple questions before, but I don't see anything related.  Is bgcolor a really recent change?

**Reply by codemasher** [reply]
> "Sure... just look through the documentation and find the custom output examples; then write some custom classes, and...."

I'm not sure how you do things, but if i'm planning to use a library, i usually go through the documentation and look for usage examples. Admittedly, the documentation could be better (in fact, i'm working on that for v5), but examples for several use cases and internal workings are provided.

> There was a reason to add the bgcolor option -- (or more specifically, the imageTransparent = false option?). It's straightforward and useful.

Honestly, i regret the `imageTransparent` option because of how [GD and Imagick handle transparency internally](https://stackoverflow.com/questions/32243/can-png-image-transparency-be-preserved-when-using-phps-gdlib-imagecopyresample). It created more confusion than it was helpful.

> Is bgcolor a really recent change?

It is so recent that it doesn't exist in v4.x, only in the v5 beta. It used to be `imageTransparencyBG` and used to work with `imagecolortransparent()` specifically. I have since changed how transparency is handled for GD and Imagick and added the more common `bgColor` which is also used in FPDF output as it behaves very similar to the aforementioned.

**Reply by strider72** [reply]
Yeah, design is subjective. Programming would be easy if it were all cut and dried.

At any rate, I realized something as I was having lunch....

>  i will not "solve" a non-issue just for the sake of consistency

This indicates a rather fundamental disagreement in terms of philosophy.  A matter of First Principles -- which means quibbling over details will never resolve with agreement.  I discussed my design philosophy above. To me, consistency in user interface is an *inherent good*. In essence, you do it simply because it's fundamentally the best way to do it, as a matter of usability -- unless there's a specific identifiable drawback that counterindicates.

You see it differently.  And in the end, that's totally cool. Different people are different.  What I mean to say is that I'm going to bow out on this one, because I don't imagine we're going to come to any agreement here.  So be it.

I'll leave with one last thing:

>  if i'm planning to use a library, i usually go through the documentation and look for usage examples

I did. I didn't look for "custom output" and the like, I looked for "background", and I saw the info on how to get a background by turning off imageTransparent and setting bgColor.  A perfect example of why (IMO) consistency is inherently good.

Peace.  It's a cool library, and I will certainly keep using it.

**Reply by codemasher** [reply]
> This indicates a rather fundamental disagreement in terms of philosophy.

It's not that there is disagreement, in fact, i try to keep things consistent *where it is possible*.
As i've stated before, there is no clean solution for this specific feature that justifies "consistency" over functionality, especially for something that can be solved in so many ways already by the built-in options (see: `svgDefs`). And i will not settle for a mediocre solution to make things consistent just for the sake of it.

Also it should be noted that several options just exist because of how the internals of several libraries/extensions work. The `imageTransparent` option exists because of GD in first place, the `bgColor` (and its former variants) because of how transparency is toggled in GD. SVG just does not need these options as it has everything built-in via CSS or attributes - and because of the different ways it can be implemented, i decided to make the SVG class as convenient to extend as possible.

**Comment by strider72** [comment]
Quick followup question: If I *do* extend the `QRMarkupSVG` class, what's the proper way to get the code to use that instead of the standard class?

EDIT: I see it. `outputType` and `outputInterface`, per https://github.com/chillerlan/php-qrcode/blob/main/examples/svgWithLogo.php

**Reply by codemasher** [reply]
Yes that's how it is supposed to work.

**Comment by shirshak007** [comment]
Can anyone help me achieve colourful QR code using OUTPUT_MARKUP_PNG
```
  // set returnResource to true to skip further processing for now
  $this->options->returnResource = true;
  
  // of course you could accept other formats too (such as resource or Imagick)
  // i'm not checking for the file type either for simplicity reasons (assuming PNG)
  if(!is_file($logo) || !is_readable($logo)){
  throw new QRCodeOutputException('invalid logo');
  }
  
  $this->matrix->setLogoSpace(
  $this->options->logoSpaceWidth,
  $this->options->logoSpaceHeight
  // not utilizing the position here
  );
  
  // there's no need to save the result of dump() into $this->image here
  parent::dump($file);
  
  $im = imagecreatefrompng($logo);
  
  // get logo image size
  $w = imagesx($im);
  $h = imagesy($im);
  
  // set new logo size, leave a border of 1 module (no proportional resize/centering)
  $lw = ($this->options->logoSpaceWidth - 2) * $this->options->scale;
  $lh = ($this->options->logoSpaceHeight - 2) * $this->options->scale;
  
  // get the qrcode size
  $ql = $this->matrix->size() * $this->options->scale;
  
  // scale the logo and copy it over. done!
  imagecopyresampled($this->image, $im, ($ql - $lw) / 2, ($ql - $lh) / 2, 0, 0, $lw, $lh, $w, $h);
  
  $imageData = $this->dumpImage();
  
  if($file !== null){
  $this->saveToFile($imageData, $file);
  } 
  
  if($this->options->imageBase64){
  $imageData = 'data:image/'.$this->options->outputType.';base64,'.base64_encode($imageData);
  }
  
  return $imageData;
```
Options: 

```
`$options = new LogoOptions(
    [
        'eccLevel' => QRCode::ECC_H,
        'imageBase64' => false,
        'logoSpaceHeight' => 17,
        'logoSpaceWidth' => 17,
        'scale' => 20,
        'version' => 7,
        'addQuietzone' => true,
        'quietzoneSize' => 1,
        
    ]
    );
```

**Reply by codemasher** [reply]
Hi, I'm not sure what exactly you're trying to achieve. Also please open a new discussion instead of replying to an unrelated one, as the OP will get unnecessary notifications.
